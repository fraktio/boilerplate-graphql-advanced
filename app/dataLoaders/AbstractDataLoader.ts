import type Knex from "knex";

export type DataLoaderParams = {
  knex: Knex;
};

export abstract class AbstractDataLoaderBase<Loaders> {
  private loaders: Loaders | null = null;

  public getLoaders(params: DataLoaderParams): Loaders {
    if (!this.loaders) {
      this.loaders = this.createLoaders({ knex: params.knex });
    }

    return this.loaders;
  }

  protected abstract createLoaders(params: DataLoaderParams): Loaders;
}
