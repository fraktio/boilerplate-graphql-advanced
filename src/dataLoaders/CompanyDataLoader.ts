import DataLoader from "dataloader";
import { DateTime } from "luxon";

import { DataLoaderParams } from "./dataLoaders";

import { CompanyTable } from "~/dataSources/CompanyDataSource";
import {
  CompanyTableRaw,
  EmployeeTableRaw,
  PersonTableRaw,
  Table,
  tableColumn,
} from "~/database/types";
import { UUID } from "~/models";

const formatRow = (row: CompanyTableRaw): CompanyTable => ({
  id: row.id,
  UUID: row.uuid,
  name: row.name,
  timestamp: {
    createdAt: DateTime.fromJSDate(row.createdAt),
    updatedAt: row.updatedAt ? DateTime.fromJSDate(row.updatedAt) : null,
  },
});

export class CompanyDataLoader {
  company: DataLoader<UUID, CompanyTable>;
  companiesOfPerson: DataLoader<UUID, UUID[]>;

  constructor(opts: DataLoaderParams) {
    this.company = new DataLoader<UUID, CompanyTable>(async (keys) => {
      const personRows = await opts
        .knex<CompanyTableRaw>(Table.COMPANY)
        .whereIn("uuid", keys);

      const persons = personRows.reduce<{ [key: string]: CompanyTable }>(
        (prev, current) => ({
          ...prev,
          [current.uuid.toString()]: formatRow(current),
        }),
        {},
      );

      return keys.map((key) => persons[key.toString()] ?? null);
    });

    this.companiesOfPerson = new DataLoader<UUID, UUID[]>(async (keys) => {
      const personsRaw = await Promise.all(
        keys.map(async (key) => {
          const companies = await opts.knex
            .select<CompanyTableRaw[]>(
              tableColumn<CompanyTableRaw>(Table.COMPANY, "uuid"),
            )
            .from(Table.COMPANY)
            .innerJoin(
              Table.EMPLOYEE,
              tableColumn<EmployeeTableRaw>(Table.EMPLOYEE, "companyId"),
              "=",
              tableColumn<CompanyTableRaw>(Table.COMPANY, "id"),
            )
            .innerJoin(
              Table.PERSONS,
              tableColumn<PersonTableRaw>(Table.PERSONS, "id"),
              "=",
              tableColumn<EmployeeTableRaw>(Table.EMPLOYEE, "personId"),
            )
            .where(tableColumn<PersonTableRaw>(Table.PERSONS, "uuid"), key);

          return {
            person: key.toString(),
            companies: companies.map((company) => company.uuid),
          };
        }),
      );

      const persons = personsRaw.reduce<{ [key: string]: UUID[] }>(
        (prev, current) => ({
          ...prev,
          [current.person]: current.companies,
        }),
        {},
      );

      return keys.map((key) => persons[key.toString()] ?? null);
    });
  }
}
