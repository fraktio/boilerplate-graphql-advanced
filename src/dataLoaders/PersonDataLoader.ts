import DataLoader from "dataloader";
import { PhoneNumberUtil } from "google-libphonenumber";
import { DateTime } from "luxon";

import { DataLoaderParams } from "./dataLoaders";

import { PersonTable } from "~/dataSources/PersonDataSource";
import {
  CompanyTableRaw,
  EmployeeTableRaw,
  PersonTableRaw,
  Table,
  tableColumn,
} from "~/database/types";
import { UUID } from "~/models";

const phoneUtil = PhoneNumberUtil.getInstance();

const formatRow = (row: PersonTableRaw): PersonTable => ({
  id: row.id,
  UUID: row.uuid,
  firstName: row.firstName,
  lastName: row.lastName,
  personalIdentityCode: row.personalIdentityCode,
  phone: phoneUtil.parseAndKeepRawInput(row.phone),
  email: row.email,
  nationality: row.nationality,
  birthday: DateTime.fromJSDate(row.birthday),
  timestamp: {
    createdAt: DateTime.fromJSDate(row.createdAt),
    updatedAt: row.updatedAt ? DateTime.fromJSDate(row.updatedAt) : null,
  },
});

export class PersonDataLoader {
  person: DataLoader<UUID, PersonTable>;
  personsOfCompany: DataLoader<UUID, UUID[]>;

  constructor(opts: DataLoaderParams) {
    this.person = new DataLoader<UUID, PersonTable>(async (keys) => {
      const personRows = await opts
        .knex<PersonTableRaw>(Table.PERSONS)
        .whereIn("uuid", keys);

      const persons = personRows.reduce<{ [key: string]: PersonTable }>(
        (prev, current) => ({
          ...prev,
          [current.uuid.toString()]: formatRow(current),
        }),
        {},
      );

      return keys.map((key) => persons[key.toString()] ?? null);
    });

    this.personsOfCompany = new DataLoader<UUID, UUID[]>(async (keys) => {
      const personsRaw = await Promise.all(
        keys.map(async (key) => {
          const persons = await opts.knex
            .select<PersonTableRaw[]>(
              tableColumn<PersonTableRaw>(Table.PERSONS, "uuid"),
            )
            .from(Table.PERSONS)
            .innerJoin(
              Table.EMPLOYEE,
              tableColumn<EmployeeTableRaw>(Table.EMPLOYEE, "personId"),
              "=",
              tableColumn<PersonTableRaw>(Table.PERSONS, "id"),
            )
            .innerJoin(
              Table.COMPANY,
              tableColumn<CompanyTableRaw>(Table.COMPANY, "id"),
              "=",
              tableColumn<EmployeeTableRaw>(Table.EMPLOYEE, "companyId"),
            )
            .where(tableColumn<CompanyTableRaw>(Table.COMPANY, "uuid"), key);

          return {
            companyKey: key.toString(),
            persons: persons.map((person) => person.uuid),
          };
        }),
      );

      const persons = personsRaw.reduce<{ [key: string]: UUID[] }>(
        (prev, current) => ({
          ...prev,
          [current.companyKey]: current.persons,
        }),
        {},
      );

      return keys.map((key) => persons[key.toString()] ?? null);
    });
  }
}
