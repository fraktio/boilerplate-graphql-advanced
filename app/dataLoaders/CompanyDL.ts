import DataLoader from "dataloader";

import { companyDB, CompanyID, CompanyTable } from "~/database/companyDB";
import { DBConnection } from "~/database/connection";

export class CompanyDL {
  dataloader: DataLoader<CompanyID, CompanyTable> | undefined;

  public getDL(params: { knex: DBConnection }) {
    if (!this.dataloader) {
      return this.createDataLoader({ knex: params.knex });
    }

    return this.dataloader;
  }

  private createDataLoader(params: { knex: DBConnection }) {
    this.dataloader = new DataLoader<CompanyID, CompanyTable>(async (keys) => {
      const personRows = await companyDB.getByIds({
        knex: params.knex,
        companyIds: keys as CompanyID[],
      });

      const persons = personRows.reduce<{ [key: string]: CompanyTable }>(
        (prev, current) => ({
          ...prev,
          [current.id.toString()]: current,
        }),
        {},
      );

      return keys.map((key) => persons[key.toString()] ?? null);
    });
  }
}
