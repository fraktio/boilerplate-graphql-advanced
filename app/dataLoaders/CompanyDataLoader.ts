import DataLoader from "dataloader";
import { DateTime } from "luxon";

import { DataLoaderParams } from "./dataLoaders";

import { CompanyTable } from "~/dataSources/CompanyDataSource";
import { CompanyTableRow, Table, tableColumn } from "~/database/utils";
import { UUID } from "~/models";

const formatRow = (row: CompanyTableRow): CompanyTable => ({
  id: row.id,
  UUID: row.uuid,
  name: row.name,
  timestamp: {
    createdAt: DateTime.fromJSDate(row.createdAt),
    updatedAt: row.updatedAt ? DateTime.fromJSDate(row.updatedAt) : null,
  },
});

type CompanyTableQueryRelationRaw = {
  companyUUID: UUID;
  personUUID: UUID;
};

export class CompanyDataLoader {
  company: DataLoader<UUID, CompanyTable>;
  companiesOfPerson: DataLoader<UUID, UUID[]>;

  constructor(opts: DataLoaderParams) {
    this.company = new DataLoader<UUID, CompanyTable>(async (keys) => {
      const personRows = await opts
        .knex<CompanyTableRow>(Table.COMPANY)
        .whereIn("uuid", keys);

      const persons = personRows.reduce<{ [key: string]: CompanyTable }>(
        (prev, current) => ({
          ...prev,
          [current.uuid.toString()]: formatRow(current),
        }),
        {},
      );

      return keys.map((key) => persons[key.toString()] ?? null);
    });

    this.companiesOfPerson = new DataLoader<UUID, UUID[]>(async (keys) => {
      const companiesRaw = await opts.knex
        .select<CompanyTableQueryRelationRaw[]>(
          `${tableColumn(Table.COMPANY, "uuid")} as companyUUID`,
          `${tableColumn(Table.PERSONS, "uuid")} as personUUID`,
        )
        .from(Table.COMPANY)
        .innerJoin(
          Table.EMPLOYEE,
          tableColumn(Table.EMPLOYEE, "companyId"),
          "=",
          tableColumn(Table.COMPANY, "id"),
        )
        .innerJoin(
          Table.PERSONS,
          tableColumn(Table.PERSONS, "id"),
          "=",
          tableColumn(Table.EMPLOYEE, "personId"),
        )
        .whereIn(tableColumn(Table.PERSONS, "uuid"), keys);

      const companies = companiesRaw.reduce<{ [key: string]: UUID[] }>(
        (prev, current) => {
          const { companyUUID } = current;
          const personUUID = current.personUUID.toString();

          return {
            ...prev,
            [personUUID]:
              personUUID in prev
                ? [...prev[personUUID], companyUUID]
                : [companyUUID],
          };
        },
        {},
      );

      return keys.map((key) => companies[key.toString()] ?? null);
    });
  }
}
