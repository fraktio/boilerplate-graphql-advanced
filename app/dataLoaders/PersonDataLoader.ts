import DataLoader from "dataloader";

import { AbstractDataLoaderBase, DataLoaderParams } from "./AbstractDataLoader";

import { personDB, PersonID, PersonTable } from "~/database/personDB";

export type PersonLoader = DataLoader<PersonID, PersonTable | null>;

export class PersonDataLoader extends AbstractDataLoaderBase<PersonLoader> {
  protected createLoaders(params: DataLoaderParams): PersonLoader {
    const personLoader: PersonLoader = new DataLoader(async (ids) => {
      const employees = await personDB.getPersonsByIds({
        knex: params.knex,
        personIds: ids,
      });

      return ids.map(
        (id) => employees.find((employee) => employee.id === id) || null,
      );
    });

    return personLoader;
  }
}
