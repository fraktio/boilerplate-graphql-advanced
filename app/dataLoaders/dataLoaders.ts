import { PersonDataLoader } from "./PersonDataLoader";

export type DataLoaders = {
  personDL: PersonDataLoader;
};

export const createDataLoaders = () => ({
  personDL: new PersonDataLoader(),
});
