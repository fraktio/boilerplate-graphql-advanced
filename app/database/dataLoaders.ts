import { CompanyDataLoader } from "./company/CompanyDataLoader";
import { CompaniesOfPersonDataLoader } from "./employee/CompaniesOfPersonDataLoader";
import { PersonsOfCompanyDataLoader } from "./employee/PersonsOfCompanyDataLoader";
import { UserDataLoader } from "./user/UserDataLoader";

import { PersonDataLoader } from "~/database/person/PersonDataLoader";

export type DataLoaders = {
  companyDL: CompanyDataLoader;
  personDL: PersonDataLoader;
  userDL: UserDataLoader;
  companiesOfPersonDL: CompaniesOfPersonDataLoader;
  personsOfCompanyDL: PersonsOfCompanyDataLoader;
};

export const createDataLoaders = (): DataLoaders => ({
  companyDL: new CompanyDataLoader(),
  personDL: new PersonDataLoader(),
  userDL: new UserDataLoader(),
  companiesOfPersonDL: new CompaniesOfPersonDataLoader(),
  personsOfCompanyDL: new PersonsOfCompanyDataLoader(),
});
