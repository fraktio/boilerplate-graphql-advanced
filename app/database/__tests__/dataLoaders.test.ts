import { createDataLoaders } from "../dataLoaders";

describe("database tests / loaders", () => {
  it("createDataLoaders", async () => {
    const loaders = createDataLoaders();

    expect(loaders).toMatchSnapshot();
  });
});
