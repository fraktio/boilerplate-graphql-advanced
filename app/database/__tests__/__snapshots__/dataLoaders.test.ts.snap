// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`database tests / loaders createDataLoaders 1`] = `
Object {
  "companyDL": CompanyDataLoader {
    "loaders": null,
  },
  "personDL": PersonDataLoader {
    "loaders": null,
  },
  "userDL": UserDataLoader {
    "loaders": null,
  },
}
`;
